### PR TITLE
Introduce new poll feature 🗳️ 

### PR DESCRIPTION
**Pull request by applenick**
This PR introduces an exciting new in-game polling feature that enables the seamless and efficient conduction of surveys. The system is designed to facilitate the effortless creation of polls, offering various options to choose from, such as command-based, set-next, player kicks, and more to come. 

Example of a custom question (can be combined with any end-action)
Screenshot 2023-08-01 at 1 37 38 AM

Example of the player kick action
Screenshot 2023-08-01 at 1 38 38 AM

Example of the set next map action
Screenshot 2023-08-01 at 1 38 02 AM

Example of poll results broadcast:
Screenshot 2023-08-01 at 1 39 35 AM

Still a WIP 🚧

### CODE REVIEW

**Comment by cswhite2000** [comment]
I'm wondering if optional custom thresholds would be a good idea. For example, for setting a time-limit or kicking a player you could have a 2/3 threshold instead of a 50% threshold for actually executing the endaction.

File: src/main/java/dev/pgm/community/polls/feature/PollFeature.java [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+  }
+
+  private void task() {
+    if (!isRunning()) return;
+
+    if (poll instanceof TimedPoll) {
+      TimedPoll timedPoll = (TimedPoll) poll;
+      Duration timeLeft = timedPoll.getTimeLeft();
+
+      if (shouldShowAlert(timeLeft)) {
```
Comment (cswhite2000) [inline src/main/java/dev/pgm/community/polls/feature/PollFeature.java]: You should consider having a different alert frequency for players who have already voted vs players who haven't voted.

Comment (applenick) [reply src/main/java/dev/pgm/community/polls/feature/PollFeature.java]: Great suggestion! I'll look into implementing different alert frequencies 👍

**Comment by applenick** [comment]
> I'm wondering if optional custom thresholds would be a good idea. For example, for setting a time-limit or kicking a player you could have a 2/3 threshold instead of a 50% threshold for actually executing the endaction.

It's indeed a good idea! Do you have any thoughts on how best to implement this? One approach could be to introduce a `/poll threshold`, followed by a disclaimer under the main poll broadcast to ensure that players are aware ahead of time. Perhaps some preset buttons under the setup message too, to make it easier for staff to select 1/2, 2/3, etc. 

On a related note, I was considering alternatives to time-based polls, such as a poll with a required player vote count.

Scenario:
`/poll question Do you want to skip this map?`
`/poll require all`
`/poll start`
In this scenario, the poll would automatically end once every online submits a vote.

However, one potential issue to address is how to handle players joining or leaving the server during the polling process. We might need to determine a fair way to draw the line when it comes to counting votes in such situations.

Brainstorming a bit further, there could be other interesting scenarios to consider, such as requiring a specific percentage of online players, just participants, observers, etc. Speaking of which, another idea just came to me – adding a vote filter to the polls for specific groups of players (participants/teams/observers) might be worth exploring too! 

Thanks @cswhite2000, for sharing your idea! If you have any further thoughts on any of these ideas, feedback would be much appreciated. 👍

**Comment by cswhite2000** [comment]
I like your suggested approach for poll thresholds 👍 

> On a related note, I was considering alternatives to time-based polls, such as a poll with a required player vote count.
> Scenario: `/poll question Do you want to skip this map?` `/poll require all` `/poll start` In this scenario, the poll would automatically end once every online submits a vote.
> However, one potential issue to address is how to handle players joining or leaving the server during the polling process. We might need to determine a fair way to draw the line when it comes to counting votes in such situations.

I don't think getting 100% of players to vote in a poll is feasible in most cases since some players will be afk or simply won't want to vote. An alternative though would be to have the option to have usage such as `/poll require 75%` and then once the threshold is reached have it automatically start a shorter timer to end the poll, to allow last-minute votes / or vote changes.

With regards to players leaving during a poll, I feel like it would be better to only count votes from players who're online at the point in time when the poll ends, as they'll be the ones actually affected by the result of the poll.

You should also consider having an automatic default expiry for such polls, to prevent them from lasting longer than their relevance.

Filtering polls for specific groups of players is a good idea.

Other ideas:
- for the command polls allow the requester to describe what the command will do
- for the mutation poll add the mutation's description to a hover
- disable mutation action/poll
- for the map poll add the map info, or the map tags to a hover
- Incentivizing voting in polls with rewards such as raindrops, to get higher participation.
    - Presumably, this would just be adding an event on Community's end
- Add a sound with alerts for players who haven't voted, possibly optional
    - could be too annoying for players
- polls with non-boolean votes
    - possibly too complicated for a poll
- map party polls ie `dev.pgm.community.party`
    - unsure of if this would have benefits over just doing a question poll.

File: src/main/java/dev/pgm/community/polls/commands/PollManagementCommands.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  }
+
+  @CommandMethod("command [command]")
+  public void command(CommandAudience audience, @Argument("command") @Greedy String command) {
+    checkPoll();
+    polls.getBuilder().command(audience, command);
+  }
+
+  @CommandMethod("mutation [mutation]")
+  public void mutation(CommandAudience audience, @Argument("mutation") MutationType mutation) {
```
Comment (cswhite2000) [inline src/main/java/dev/pgm/community/polls/commands/PollManagementCommands.java:144]: Presumably, this should check the mutation permission here, as someone could conceivably have poll perms but not mutation perms, which would let them make a confusing poll to add a mutation, which wouldn't actually do so.

File: src/main/java/dev/pgm/community/polls/commands/PollVoteCommands.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+public class PollVoteCommands extends CommunityCommand {
+
+  private final PollFeature polls;
+
+  public PollVoteCommands() {
+    this.polls = Community.get().getFeatures().getPolls();
+  }
+
+  @CommandMethod("yes")
```
Comment (Pugzy) [inline src/main/java/dev/pgm/community/polls/commands/PollVoteCommands.java]: I think these should be prefixed with something, maybe taking inspiration from the vote book.
`votenext -o [option]` so maybe a `votepoll -o [option]`

The commands for these actions are not expected to be typed by players but from clickable text so there's no need to make them so generic which could cause other issues.

Comment (Pablete1234) [reply src/main/java/dev/pgm/community/polls/commands/PollVoteCommands.java]: note, -o is for "open", not for the option, so just `/votepoll [option]`

Comment (applenick) [reply src/main/java/dev/pgm/community/polls/commands/PollVoteCommands.java]: Great suggestion 👍 Will help pave the way for the eventual multi-answer polls.

Comment (Pugzy) [reply src/main/java/dev/pgm/community/polls/commands/PollVoteCommands.java]: > note, -o is for "open", not for the option, so just `/votepoll [option]`

🤦 I added the `-o` flag 😆 it was 3 years ago so please forgive me.

**Comment by applenick** [comment]
I'm beginning to explore how to implement multi-answer polls. Does anyone have any thoughts on the most effective way to implement them?

Ideally was thinking to have a maximum of 4 options, and simply executing multiple end-actions commands will be how staff can add multiple options to a poll. When no custom question is defined we'll auto-generate one based on the options. Here's a few examples to illustrate it:

Example: all mutations
`/poll mutation rage`
`/poll mutation bread`
`/poll mutation mobs`
`/poll mutation blitz`
`/poll start`

```
----------------- Server Poll -----------------
  Question: Which mutation should we enable?

	[Rage] [Bread] [Mobs] [Blitz]

-----------------------------------------------
```

Example: 2 different maps
`/poll map Race for Victory`
`/poll map Directrix`
`/poll start`


```
----------------- Server Poll -----------------
  Question: Which map should we play next?

	[Race for Victory] [Directrix] 

-----------------------------------------------
```


Though let's think of a more complex example. We have a 3 hour match going and staff wants to spice things up.

Example: multiple end actions
`/poll command /spawn wither 100`
`/poll mutation fly`
`/poll kick applenick`
`/poll start`

```
----------------- Server Poll -----------------
  	Question: What should we do?

[/spawn wither 100] [Fly Mutation] [Kick applenick] 

-----------------------------------------------
```
When multiple end-actions are included in a poll, the question will be more general (unless a custom question is defined). Depending on the type, the button will also feature additional text to provide clarity about the option.
Command -> display the command
Mutation -> `<name> Mutation`
Map -> `Set <name>`
Kick -> `Kick <name>`

If anyone has further thoughts or ideas, please feel free to share! I won't be starting on this right away, but I wanted to brainstorm beforehand to ensure we keep a clear implementation and simple command structure in place for staff.

**Comment by applenick** [comment]
Polls on OCC have remained stable for some time now. As just about all suggested features have been implemented, going to merge now! 🎉

**Pull request merged**